### ISSUE TITLE
device cert auth should account for SCEP renewal

### ISSUE THREAD

**Issue by groob** [micromdm/micromdm#557]

**Comment by jessepeterson** [comment]
This _should_ happen if we get a new Authenticate Checkin message after the SCEP renewal, but I'll double-check.

**Comment by bdemetris** [comment]
hey @jessepeterson have you had a chance to look at this at all?

ive been doing a bit of digging myself, and it seems like the auth middleware is leaning on the `Authenticate` message to unconditionally save the certhash.  according to the mdm spec that message only happens after an mdm payload is installed.  however, it doesn't look like the scep renewal is an mdm payload (at least that is what i reason), because if i renew a cert i dont see any mdm messages associated with the transaction.  then, on the next actual mdm transaction (i.e i just push a device info command) im getting `level=info component=udidcertauth msg="device cert hash mismatch"`

**Comment by jessepeterson** [comment]
@bdemetris as far as I can tell we're operating to the spec. As noted in #358 and per the spec we're supposed to validate the certificate against the UDID:

> An MDM server should identify a connecting device by examining the deviceʼs client identity certificate. The server should then cross-check the UDID reported in the message to ensure that the UDID is associated with the certificate.

As you note it appears Apple has no supported way to inform the MDM of an updated certificate that is done outside of the MDM protocol (i.e. the SCEP payload renewal). It appears the only supported way to "renew" the MDM identity certificate is to re-send the enrollment profile via MDM which re-enrolls the device and sends a new `Authenticate` message.

MDM SCEP certificates cannot be automatically renewed. [HT204836](https://support.apple.com/en-us/HT204836) states that you can optionally turn on auto-renewal for SCEP and AD payloads but when doing this the MDM SCEP payloads are expressly disallowed per the Apple logs. Even though the tech note specifically states it's supposed to work. So it appears the only way to get an MDM SCEP payload renewed is manually — via the GUI or with `profiles -W`. Which calls into question the value of even considering supporting MDM SCEP payload renewal anyway and just relying on the above re-enrollment. We aren't validating device certificate expiry so expired device identity certificates of MDM clients continue to work just fine. In fact, as you notice, manually renewed clients will indeed stop working!

One interesting side note is that if you're enrolled via OTA profiles on macOS when you trigger the SCEP renewal the device actually re-enrolls itself just as if you sent it the MDM enrollment profile via MDM, thus getting a new identity certificate, sending an `Authenticate` message and properly associating the UDID. This doesn't appear to happen with non-OTA enrollments nor DEP enrollments. I suspect this is because the certificate that is shown to be expiring is the OTA SCEP exchange, not the MDM SCEP certificate.

In general I'd be against somehow modifying the SCEP server to be "smart" about recognizing which SCEP renewal belongs to which client and reaching back into the UDID-Cert auth database to update that. For a couple reasons: 1) we shouldn't assume any knowledge or interaction with the SCEP server (to support those configurations where folks want/need to use other SCEP servers) 2) it just feels absolutely hacky in general to try and attempt that. I'd much rather try and lobby Apple to adopt some sort of `ReAuthenticate` message 3) I'm not sure that the SCEP server currently validates the private key of a renewal request based on an existing cert (or if the client generates a new private key); i.e. the security of a SCEP renewal itself hasn't been vetted that I know of). 

So a couple things that can be done:

1. Re-send the MDM enrollment profile (via MDM) to clients that are or are about to expire. This will require keeping track of enrollment date (server-side or client-side).
   * Other MDM vendors appear to use this strategy, or a _very_ long device identity certificate timeline (e.g. 10 years).
1. MicroMDM could change the "device cert hash mismatch" error to be a soft error/warning, instead of a failure.
1. (Per above, what I _don't_ want to do) modify the SCEP server to know about UDIDs and be "smart" about certificate renewals.

**Comment by bdemetris** [comment]
hey @jessepeterson sounds like you have been thinking about this!

I think 1 is a good idea, but I also think it raises the bar a bit for adoption of micromdm.  Like, you must be this tall to ride micromdm (i.e you need to develop tooling outside of the platform to use micromdm successfully).

The second option sounds like a regression.  What if instead, on a failure to validate you just que a profile push to the device for the admin?  The user gets kind of screwed this way, since 14 days prior their machine told them to do something mysterious (and they clicked renew), and then you're going to push them a second mysterious thing that's asking for a lot of device permissions.  At least with this approach, the admin still has a low barrier to entry for using micromdm successfully, and if they want their users to be happier they would build the tooling you described for option 1.

The 3rd option, while it sounds terrible, is probably the easiest. (just ignore it and make certificates not matter since thats basically how apple set it up, and we dont have a `ReAuthenticate` message), and sounds like all of the others do it this way (is everyone wrong? /shrug)

would you be willing to take a PR for making the hard coded 365 day certificate validity https://github.com/micromdm/micromdm/blob/master/server/server.go#L338-L341 an option for the server at startup?  I dont think this closes the bug, but it does give admins a choice (maybe some will even pick a shorter validity?)

**Comment by jessepeterson** [comment]
> I think 1 is a good idea, but I also think it raises the bar a bit for adoption of micromdm.

That's true, but I'm not sure keeping the bar low is MicroMDM's target. :)

> What if instead, on a failure to validate you just que a profile push to the device for the admin? [...]

Once the device fails to validate then you'd have to turn off validation in order to queue anything to it, no? I agree it's not good when this happens — can leave devices (and users) hanging.

> The user gets kind of screwed this way, since 14 days prior their machine told them to do something mysterious (and they clicked renew), and then you're going to push them a second mysterious thing that's asking for a lot of device permissions.

Not sure I follow. If we just turned the error into a warning then the user gets nothing (after their manual renewal step). It just keeps working no matter what they do or don't do. Still not ideal, though.

> The 3rd option, while it sounds terrible, is probably the easiest. (just ignore it and make certificates not matter since thats basically how apple set it up, and we dont have a `ReAuthenticate` message), and sounds like all of the others do it this way (is everyone wrong? /shrug)

Do you mean the 2nd option here? Simply warn instead of fail? The 3rd option I wrote about was to reverse-track the UDID from the SCEP certificate renewal in some (hopefully) secure way. But that option just feels moot anyway because it can't be automated. Anything my entire fleet of users has to take manual steps to do for me is sort of a non-starter. :)

> would you be willing to take a PR for making the hard coded 365 day certificate validity

Not only would I be willing to accept that (great suggestion), but I'd also consider just changing the default certificate timeline to 10 years (or whatever) as a part of that. Longer expiry seems to be industry standard, it would allow us to keep the validation, avoid the "smart" SCEP server plan, have a sane default for folks (absent any other logic/workflows), and allow for further testing (e.g. setting it to 1 day, etc.). Unfortunate Apple didn't think this through, but it also is what it is.

**Comment by jessepeterson** [comment]
This is summary of the situation to anyone joining the party: https://github.com/micromdm/micromdm/wiki/Device-Identity-Certificate-Expiration

**Comment by bdemetris** [comment]
@jessepeterson as we discussed https://github.com/micromdm/micromdm/pull/577 will allow admins to set a desired validity in days.

I wonder if this does resolve the bug or we'd just treat it as a known issue?
